Add example usage to DataFrame.filter

cswarth opened the pull request:
Updates doc comments for `DataFrame.filter` and adds usage examples.
Fixed errors identified by `flake8` and correctly rebase my branch before issuing PR.

<dl class="method">
<dt id="pandas.DataFrame.filter">
<code class="descclassname">DataFrame.</code><code class="descname">filter</code><span class="sig-paren">(</span><em>items=None</em>, <em>like=None</em>, <em>regex=None</em>, <em>axis=None</em><span class="sig-paren">)</span><a class="headerlink" href="#pandas.DataFrame.filter" title="Permalink to this definition">¶</a></dt>
<dd><p>Subset rows or columns of dataframe according to  labels in the index.</p>
<p>Note that this routine does not filter a dataframe on its contents. The filter is 
applied to the labels of the index.
This method is a thin veneer on top of <span class="xref std std-ref">DateFrame Select</span></p>
<table class="docutils field-list" frame="void" rules="none">
<colgroup><col class="field-name">
<col class="field-body">
</colgroup><tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><p class="first"><strong>items</strong> : list-like</p>
<blockquote>
<div><p>List of info axis to restrict to (must not all be present)</p>
</div></blockquote>
<p><strong>like</strong> : string</p>
<blockquote>
<div><p>Keep info axis where “arg in col == True”</p>
</div></blockquote>
<p><strong>regex</strong> : string (regular expression)</p>
<blockquote>
<div><p>Keep info axis with re.search(regex, col) == True</p>
</div></blockquote>
<p><strong>axis</strong> : int or None</p>
<blockquote>
<div><p>The axis to filter on.</p>
</div></blockquote>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first last">same type as input object with filtered info axis</p>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Notes</p>
<p>The <code class="docutils literal"><span class="pre">items</span></code>, <code class="docutils literal"><span class="pre">like</span></code>, and <code class="docutils literal"><span class="pre">regex</span></code> parameters should be mutually exclusive, but this is not checked.</p>
<p><code class="docutils literal"><span class="pre">axis</span></code> defaults to the info axis that is used when indexing with <code class="docutils literal"><span class="pre">[]</span></code>.</p>
<p class="rubric">Examples</p>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="n">df</span>
<span class="go">        one  two  three</span>
<span class="go">mouse     1    2      3</span>
<span class="go">rabbit    4    5      6</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="c1"># select columns by name</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">items</span><span class="o">=</span><span class="p">[</span><span class="s1">'one'</span><span class="p">,</span> <span class="s1">'three'</span><span class="p">])</span>  
<span class="go">        one  three</span>
<span class="go">mouse     1      3</span>
<span class="go">rabbit    4      6</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="c1"># select columns by regular expression</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">regex</span><span class="o">=</span><span class="s1">'e$'</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="go">        one  three</span>
<span class="go">mouse     1      3</span>
<span class="go">rabbit    4      6</span>
</pre></div>
</div>
<div class="highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="c1"># select rows containing 'bbi'</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">df</span><span class="o">.</span><span class="n">filter</span><span class="p">(</span><span class="n">like</span><span class="o">=</span><span class="s1">'bbi'</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
<span class="go">        one  two  three</span>
<span class="go">rabbit    4    5      6</span>
</pre></div>
</div>
</dd></dl>

cswarth committed:
DOC: Add example usage to DataFrame.filter

- add sample usage
- enforce mutual exlusion of keyword arguments
- add note to what's new doc calling out API change

jreback commented on pandas/core/generic.py:
```diff
@@ -2312,8 +2312,15 @@ def _reindex_axis(self, new_index, fill_method, axis, copy):
+        Note that this routine does not filter a dataframe on its
+        contents. The filter is applied to the labels of the index.
+        This method is a thin veneer on top of :ref:`DateFrame Select
+        <DataFrame.select>`
```
this last sentence is not necessary.

jreback commented on pandas/core/generic.py:
```diff
@@ -2324,15 +2331,44 @@ def filter(self, items=None, like=None, regex=None, axis=None):
         regex : string (regular expression)
             Keep info axis with re.search(regex, col) == True
         axis : int or None
```
this axis args should be generated like we do for many other methods, see ``.reindex`` or ``.fillna``

jreback replied:
can you make this change

jreback replied:
can you update the axis parameters as well (can be a string)

jreback commented on pandas/core/generic.py:
```diff
@@ -2324,15 +2331,44 @@ def filter(self, items=None, like=None, regex=None, axis=None):
         Notes
         -----
-        Arguments are mutually exclusive, but this is not checked for
+        The ``items``, ``like``, and ``regex`` parameters should be
```
this is not checked? are you sure? if so then that's a much bigger problem. defering to documentation is not a good idea.

cswarth replied:
Yes, mutual exclusion of parameters really is not checked.   That line was already in the docs, I think I just reworded it a bit.  As @MaximilianR said of this routine, "[it seems to do a lot of stuff not very well](https://github.com/pydata/pandas/pull/12395#issuecomment-186368823)"

jreback replied:
can you put in a check for this and a couple of tests?

jreback commented on pandas/core/generic.py:
```diff
@@ -2324,15 +2331,44 @@ def filter(self, items=None, like=None, regex=None, axis=None):
-        Arguments are mutually exclusive, but this is not checked for
+        The ``items``, ``like``, and ``regex`` parameters should be
+        mutually exclusive, but this is not checked.
 
```
you can add a see also section to direct to ``.select``

cswarth commented:
Removed the "thin veneer" comment and added a See Also referring to pandas.DataFrame.select.
The point is to demonstrate that this routine does not filter on the contents of the dataframe, but on the index.  Hopefully this will help those coming from R who might be expecting `DataFrame.filter` to act like `dplyr::filter`

jreback commented:
can you update

cswarth commented:
Sorry for being a newbie, but update in what way?  Do you want me to rebase my `doc/df_filter` development branch to `upstream/master` and push that to github?  It looks like that will be push 201 commits so I want to make sure that is the right thing to do.

jreback commented:
you need to rebase on master first

cswarth commented:
I updated my branch, but 24hrs later the CI builds are still not complete.  Is that unusual or expected?

Also the CI build for python 3.5 failed but I don't see how it has anything to do with changes I created.
https://travis-ci.org/pydata/pandas/jobs/124031500

jreback commented:
Travis was having some issues ok now

ithe 3.5 build on. numpy master is currently failing but that is ok

jreback commented:
can you rebase / update according to comments

jreback commented:
can you rebase / update

jreback commented:
can you update

cswarth commented:
As requested, added test for mutually exclusive arguments in DataFrame.filter and added unit tests for same.

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -2368,18 +2371,58 @@ def filter(self, items=None, like=None, regex=None, axis=None):
         import re
 
+        args = locals().copy()
+        nkw = sum(map(lambda x: operator.getitem(args,x)!=None, ['items', 'like', 'regex']))
```
This seems a bit complicated for what we want to achieve (the locals is certainly not needed I think). 
Using a similar construct: `sum(map(lambda x: x is not None, [item, like, regex]))` (but using a list comprehension is maybe easier to read: `sum([x is not None for x in [item, like, regex]])`)

cswarth replied:
thanks, changing to,
```
nkw = sum([operator.getitem(args,x) is not None for x in ['items', 'like', 'regex']])
if nkw > 1:
        raise TypeError("filter(): keyword arguments are mutually exclusive")
```

cswarth replied:
opps, I didn't understand...now using,

jorisvandenbossche commented on pandas/core/generic.py:
```diff
@@ -2368,18 +2371,58 @@ def filter(self, items=None, like=None, regex=None, axis=None):
+        args = locals().copy()
+        nkw = sum(map(lambda x: operator.getitem(args,x)!=None, ['items', 'like', 'regex']))
+        if nkw == 0:
+            raise TypeError("filter(): must specify at least one keyword argument")
```
This one is already raised below

cswarth replied:
deleted this exception and updated the tests to expect the latter one.

jorisvandenbossche commented on pandas/tests/test_generic.py:
```diff
@@ -1497,6 +1497,40 @@ def test_to_xarray(self):
                            expected,
                            check_index_type=False)
 
+    def test_filter(self):
```
There is already a `test_filter` in https://github.com/pydata/pandas/blob/c2ea8fb20611370c7db7e11506727b0b67a76662/pandas/tests/frame/test_axis_select_reindex.py#L621, maybe good to add this there?

cswarth replied:
Thanks, I'll move the tests there.

cswarth commented:
Enforcing mutually exclusive arguments in frame.filter() is an incompatible change that is going to break some code, somewhere.  Should this be called out in the release notes?

I would like to make a case that this API should be deprecated at this time.   It adds little value over reindex and select at the expense of an API that is different than just about anything else.

jreback commented on pandas/core/generic.py:
```diff
@@ -2368,18 +2371,56 @@ def filter(self, items=None, like=None, regex=None, axis=None):
+
         if axis is None:
             axis = self._info_axis_name
         axis_name = self._get_axis_name(axis)
```
e.g. checked here

jreback commented:
@cswarth yes pls add a whatsnew note about the mutually exclusive args.

jreback commented:
already slated to deprecate in 0.19.0: https://github.com/pydata/pandas/issues/12401

jreback commented:
thanks @cswarth
